src/Features/Core/Diagnostics/DiagnosticData.cs

```diff
@@ -284,6 +284,8 @@ private static void SwapIfNeeded(ref LinePosition startLinePosition, ref LinePos
 
         public static DiagnosticData Create(Workspace workspace, Diagnostic diagnostic)
         {
+            Contract.Requires(diagnostic.Location == null || !diagnostic.Location.IsInSource);
```

Comment:
Diagnostic has internal abstract methods so all implementations should be internal. I don't think we ever want to allow public implementations of Diagnostic, that will probably cause lots of other unwanted issues. We can probably add an assert in our ReportDiagnostic methods that it is of one of our internal Diagnostic types.